<a href="https://colab.research.google.com/github/Hevej/Machine-Learning/blob/main/Nivel_Educativo/NivelEducativo_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Conexión con Drive
Para poder acceder a nuestra unidad en la nube y guardar o leer archivos desde alli.
Tener cuidado al leer directamente, puede generar problemas. Se recomienda leer desde .zip y descomprimir en la sesión, mas no leer directamente.

Se puede extraer información en local mediante el comando en consola:


```
!unzip -qq "/content/Train.zip"
```

En donde "Train.zip" sera nuestro archivo

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# Importar la librerias
- Pandas : Exploración y manipulación de datos


In [ ]:
import pandas as pd

#Leer el dataset

En este caso, al ser archivos no tan pesados, podemos leer directamente del drive, para esto, primero guardamos la ruta del archivo en el drive para cada base de datos.

In [ ]:
FOLDER_PATH = '/content/drive/MyDrive/SisbenColombia'
HOGAR = '/content/drive/MyDrive/SisbenColombia/sisben_hog_siv_2021.sas7bdat'
PERSONA = '/content/drive/MyDrive/SisbenColombia/sisben_pers_siv_2021.sas7bdat'
VIVIENDA = '/content/drive/MyDrive/SisbenColombia/sisben_viv_siv_2021.sas7bdat'

La parte esencial de Pandas es el DataFrame, es como guardar la información en forma de tablas, parecido a Excel o a SQL. Accedemos a las bases de datos con Pandas y la libreria las guardara como un Dataframe.

In [ ]:
dataHogar = pd.read_sas(HOGAR)
dataPersona = pd.read_sas(PERSONA)
dataVivienda =pd.read_sas(VIVIENDA)

# Exploración Inicial

Una vez en la forma de DataFrame, podemos hacer una exploración inicial.

In [ ]:
#Para este caso no tiene mucho sentido obtener estos estadisticos, debido a que hay datos de todas las naturalezas, numericos y categoricos.
dataHogar.describe()

cod_mpio          ZONA         llave         HOGAR           FEX  \
count  1.341695e+06  1.341695e+06  1.341695e+06  1.341695e+06  1.341695e+06   
mean   3.848666e+04  1.518041e+00  6.100421e+02  1.199688e+00  6.466546e+00   
std    2.529996e+04  4.996746e-01  4.275133e+02  5.633826e-01  4.153483e+01   
min    5.001000e+03  1.000000e+00  1.000000e+00  1.000000e+00  1.034483e+00   
25%    1.584200e+04  1.000000e+00  2.670000e+02  1.000000e+00  1.927500e+00   
50%    2.709900e+04  2.000000e+00  5.420000e+02  1.000000e+00  2.844584e+00   
75%    5.482000e+04  2.000000e+00  8.760000e+02  1.000000e+00  4.655072e+00   
max    9.977300e+04  2.000000e+00  3.057000e+03  1.500000e+01  1.747659e+03   

             HOG001        HOG002        HOG003        HOG004        HOG005  \
count  1.341695e+06  1.341695e+06  1.341695e+06  1.341695e+06  1.341695e+06   
mean   3.043348e+00  2.055433e+00  1.511786e+00  1.988448e+00  1.844325e+00   
std    1.196062e+00  1.188913e+00  8.427762e-01  1.155367e+00  1.988089e+00   
min    1.000000e+00  0.000000e+00  0.000000e+00  1.000000e+00  1.000000e+00   
25%    3.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
50%    3.000000e+00  2.000000e+00  1.000000e+00  2.000000e+00  1.000000e+00   
75%    4.000000e+00  3.000000e+00  2.000000e+00  2.000000e+00  2.000000e+00   
max    5.000000e+00  1.500000e+01  1.500000e+01  5.000000e+00  9.000000e+00   

       ...        HOG018        HOG019        HOG020        HOG021  \
count  ...  1.341695e+06  1.341695e+06  1.341695e+06  1.341695e+06   
mean   ...  1.568285e+00  1.824273e+00  1.951992e+00  1.858401e+00   
std    ...  4.953154e-01  3.805878e-01  2.137831e-01  3.486380e-01   
min    ...  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    ...  1.000000e+00  2.000000e+00  2.000000e+00  2.000000e+00   
50%    ...  2.000000e+00  2.000000e+00  2.000000e+00  2.000000e+00   
75%    ...  2.000000e+00  2.000000e+00  2.000000e+00  2.000000e+00   
max    ...  2.000000e+00  2.000000e+00  2.000000e+00  2.000000e+00   

             HOG022        HOG023        HOG024        HOG025        HOG026  \
count  1.341695e+06  1.341695e+06  1.341695e+06  1.341695e+06  1.341695e+06   
mean   1.959153e+00  1.992837e+00  1.992194e+00  1.900339e+00  1.954104e+00   
std    1.979354e-01  8.433279e-02  8.800496e-02  2.995479e-01  2.092605e-01   
min    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    2.000000e+00  2.000000e+00  2.000000e+00  2.000000e+00  2.000000e+00   
50%    2.000000e+00  2.000000e+00  2.000000e+00  2.000000e+00  2.000000e+00   
75%    2.000000e+00  2.000000e+00  2.000000e+00  2.000000e+00  2.000000e+00   
max    2.000000e+00  2.000000e+00  2.000000e+00  2.000000e+00  2.000000e+00   

             HOG027  
count  1.341695e+06  
mean   2.762403e+00  
std    1.577284e+00  
min    1.000000e+00  
25%    1.000000e+00  
50%    3.000000e+00  
75%    4.000000e+00  
max    1.900000e+01  

[8 rows x 32 columns]

Se generan otras exploraciones previas que entreguen mas información

Nombre de las columnas

In [ ]:
#Imprimir el nombre de las columnas nos servira para saber que features tenemos disponibles en nuestro dataset
print(dataHogar.columns)
print(dataPersona.columns)
print(dataVivienda.columns)

Index(['cod_mpio', 'ZONA', 'llave', 'CORTE', 'HOGAR', 'FEX', 'HOG001',
       'HOG002', 'HOG003', 'HOG004', 'HOG005', 'HOG006', 'HOG007', 'HOG008',
       'HOG009', 'HOG010', 'HOG011', 'HOG012', 'HOG013', 'HOG014', 'HOG015',
       'HOG016', 'HOG017', 'HOG018', 'HOG019', 'HOG020', 'HOG021', 'HOG022',
       'HOG023', 'HOG024', 'HOG025', 'HOG026', 'HOG027'],
      dtype='object')
Index(['cod_mpio', 'H_5', 'I1', 'I2', 'I3', 'I4', 'I5', 'I6', 'I7', 'I8', 'I9',
       'I10', 'I11', 'I12', 'I13', 'I14', 'I15', 'Grupo', 'Nivel', 'ZONA',
       'llave', 'CORTE', 'HOGAR', 'ORDEN', 'FEX', 'PER001', 'PER002', 'PER003',
       'PER004', 'PER005', 'PER005B', 'PER006', 'PER006B', 'PER007', 'PER008',
       'PER009', 'PER010', 'PER011', 'PER012', 'PER013', 'PER014', 'PER015',
       'PER016', 'PER017', 'PER018', 'PER019', 'PER020', 'CLASIFICACION'],
      dtype='object')
Index(['cod_mpio', 'ZONA', 'llave', 'CORTE', 'FEX', 'VIV001', 'VIV002',
       'VIV003', 'VIV004', 'VIV005', 'VIV006', 'VIV007', '

Cantidad de registros por base de datos

In [ ]:
print('Hogar:',len(dataHogar),'Persona:', len(dataPersona),'Vivienda:',len(dataVivienda))

Hogar: 1341695 Persona: 3706298 Vivienda: 1148127


Tipo de datos para cada columna

In [ ]:
print(dataHogar.dtypes)
print(dataPersona.dtypes)
print(dataVivienda.dtypes)

cod_mpio    float64
ZONA        float64
llave       float64
CORTE        object
HOGAR       float64
FEX         float64
HOG001      float64
HOG002      float64
HOG003      float64
HOG004      float64
HOG005      float64
HOG006      float64
HOG007      float64
HOG008      float64
HOG009      float64
HOG010      float64
HOG011      float64
HOG012      float64
HOG013      float64
HOG014      float64
HOG015      float64
HOG016      float64
HOG017      float64
HOG018      float64
HOG019      float64
HOG020      float64
HOG021      float64
HOG022      float64
HOG023      float64
HOG024      float64
HOG025      float64
HOG026      float64
HOG027      float64
dtype: object
cod_mpio         float64
H_5              float64
I1               float64
I2               float64
I3               float64
I4               float64
I5               float64
I6               float64
I7               float64
I8               float64
I9               float64
I10              float64
I11              float64
I

#Merge del dataset
Al tener varias bases de datos juntas, es necesario juntarlas segun las llaves foraneas o Foreign Key (FK), ajustar la cantidad de registro si es necesario.

Hacemos un merge por las llaves foraneas: 
> 'cod_mpio','ZONA','llave','CORTE','HOGAR'.

Este merge se hara entre "dataHogar" y "dataPersona"



In [ ]:
dataHogarPersona = pd.merge(dataHogar,dataPersona, on=['cod_mpio','ZONA','llave','CORTE','HOGAR'])

Haremos lo mismo con "dataVivienda" y la adicionaremos ak DataFrame ya creado anteriormente, "dataHogarPersona". Esto se hace mediante las llaves foraneas:
> 'cod_mpio', 'ZONA', 'llave', 'CORTE'



In [ ]:
data = pd.merge(dataHogarPersona,dataVivienda, on=['cod_mpio','ZONA','llave','CORTE'])

Verificamos que las columnas hayan quedado correctamente y que la longuitud del DataFrame corresponda.

In [ ]:
print(data.columns)
print(len(data))

Index(['cod_mpio', 'ZONA', 'llave', 'CORTE', 'HOGAR', 'FEX_x', 'HOG001',
       'HOG002', 'HOG003', 'HOG004', 'HOG005', 'HOG006', 'HOG007', 'HOG008',
       'HOG009', 'HOG010', 'HOG011', 'HOG012', 'HOG013', 'HOG014', 'HOG015',
       'HOG016', 'HOG017', 'HOG018', 'HOG019', 'HOG020', 'HOG021', 'HOG022',
       'HOG023', 'HOG024', 'HOG025', 'HOG026', 'HOG027', 'H_5', 'I1', 'I2',
       'I3', 'I4', 'I5', 'I6', 'I7', 'I8', 'I9', 'I10', 'I11', 'I12', 'I13',
       'I14', 'I15', 'Grupo', 'Nivel', 'ORDEN', 'FEX_y', 'PER001', 'PER002',
       'PER003', 'PER004', 'PER005', 'PER005B', 'PER006', 'PER006B', 'PER007',
       'PER008', 'PER009', 'PER010', 'PER011', 'PER012', 'PER013', 'PER014',
       'PER015', 'PER016', 'PER017', 'PER018', 'PER019', 'PER020',
       'CLASIFICACION', 'FEX', 'VIV001', 'VIV002', 'VIV003', 'VIV004',
       'VIV005', 'VIV006', 'VIV007', 'VIV008', 'VIV009', 'VIV010'],
      dtype='object')
3706298


Si recordamos:
> Hogar: 1341695 Persona: 3706298 Vivienda: 1148127

Esas eran las bases de datos por su cantidad de registros, eso significa que el anterior DataFrame creado "data" posee valores nulos que deberemos corregir o que para un hogar, se han contado varias personas. 

Para esto, se prueba si alguna fila es nulo:

In [ ]:
nulos = data.isnull().sum()
nulos.argmax()

0

No tenemos ningun registro nulo, por lo que hay personas que compartiran información de la base de datos de Hogar y de Vivienda.

Si ese no fuera el caso y tuvieramos data nula, como no poseemos data para rellenar esos vacios, la decisión hhubiera sido eliminar los registros que no posean todos los campos.

Otras opciones para trabajar con datos vacios serian:

- ``isnull()``: Para generar una mascara sobre los datos nulos
- ``notnull()``: Los contrario a ``isnull()``
- ``fillna()``: Retornar una copia del DataFrame con los datos nulos reemplazados por un valor

---

Esta seria la linea correcta

```
data.dropna(how='any', inplace=True)
```

pero debido al consumo tan grande que requiere hacer esta operación (y otras mas), haremos una reducción de memoria mediante algunas tecnias

# Reducción de memoria

Cambiar los tipos de datos por otros mas pequeños
> Esto se puede hacer con ayuda de la descripción del dataset, que contiene la naturaleza de cada variable.

Mediante el siguiente codigo



```
df['A'] = df['A'].astype(int)
```



Según la variables que tengo, las organizo en los tipos correctos:



```
['cod_mpio', 'ZONA', 'llave', 'CORTE', 'HOGAR', 'FEX_x', 'HOG001',
       'HOG002', 'HOG003', 'HOG004', 'HOG005', 'HOG006', 'HOG007', 'HOG008',
       'HOG009', 'HOG010', 'HOG011', 'HOG012', 'HOG013', 'HOG014', 'HOG015',
       'HOG016', 'HOG017', 'HOG018', 'HOG019', 'HOG020', 'HOG021', 'HOG022',
       'HOG023', 'HOG024', 'HOG025', 'HOG026', 'HOG027', 'H_5', 'I1', 'I2',
       'I3', 'I4', 'I5', 'I6', 'I7', 'I8', 'I9', 'I10', 'I11', 'I12', 'I13',
       'I14', 'I15', 'Grupo', 'Nivel', 'ORDEN', 'FEX_y', 'PER001', 'PER002',
       'PER003', 'PER004', 'PER005', 'PER005B', 'PER006', 'PER006B', 'PER007',
       'PER008', 'PER009', 'PER010', 'PER011', 'PER012', 'PER013', 'PER014',
       'PER015', 'PER016', 'PER017', 'PER018', 'PER019', 'PER020',
       'CLASIFICACION', 'FEX', 'VIV001', 'VIV002', 'VIV003', 'VIV004',
       'VIV005', 'VIV006', 'VIV007', 'VIV008', 'VIV009', 'VIV010']
```

In [ ]:
 #Todos los int
enteros = ['cod_mpio','llave', 'VIV009', 'VIV010', 'HOG002', 'HOG003',
           'HOG009', 'HOG011', 'HOG027', 'PER005B', 'PER006B']
for col in enteros:
  data[col] = data[col].astype(int)

In [ ]:
#Todas las categorias
categorias = ['ZONA','VIV001','VIV002', 'VIV003', 'VIV004', 'VIV005',
              'VIV006','VIV007', 'VIV008', 'HOGAR', 'HOG001', 'HOG004',
              'HOG005','HOG006', 'HOG007', 'HOG008', 'HOG010', 'HOG012',
              'HOG013', 'HOG014', 'HOG015', 'HOG016', 'HOG017', 'HOG018',
              'HOG019', 'HOG020', 'HOG021', 'HOG022', 'HOG023', 'HOG024',
              'HOG025', 'HOG026', 'H_5', 'I1', 'I2', 'I3', 'I4', 'I5', 'I6',
              'I7', 'I8', 'I9', 'I10', 'I11', 'I12', 'I13', 'I14', 'I15',
              'Nivel', 'ORDEN', 'PER001', 'PER002', 'PER003', 'PER004', 'PER005',
              'PER006', 'PER007', 'PER008', 'PER009', 'PER010', 'PER011', 'PER012',
              'PER013', 'PER014', 'PER015', 'PER016', 'PER017', 'PER018', 'PER019',
              'PER020', 'CLASIFICACION']
for col in categorias:
  data[col] = data[col].astype('category')

Algunas generalidades extra de los datos

String:

  CORTE, Grupo

float64:

  FEX_x, FEX_y, FEX

Irrelevante:

  CORTE, ORDEN, llave, FEX_x, FEX_y, PER005B, PER006B, PER008, FEX

Variable de interes:

  PER017

Corroboramos que haya quedado con los nuevos tipos

In [ ]:
data.dtypes

cod_mpio       int64
ZONA        category
llave          int64
CORTE         object
HOGAR       category
              ...   
VIV006      category
VIV007      category
VIV008      category
VIV009         int64
VIV010         int64
Length: 87, dtype: object

En este punto, hemos usado mucha memoria al guardar todos los DataFrame, por lo que eliminaremos las columnas irrelevantes y guardaremos el DataFrame en un .CSV para poder exportar el trabajo que llevamos hasta este punto.

# Eliminación de columnas irrelevantes

In [ ]:
data = data.drop(['CORTE', 'ORDEN', 'llave', 'FEX_x', 'FEX_y', 'PER005B', 'PER006B', 'PER008', 'FEX'], axis=1)

In [ ]:
len(data.columns)

78

# Guardar el DataFrame

Aunque por lo general uno guarda la información en un CSV, lo ideal es guardarlo en otro formato, por dos razones:

* Es muy lento y consume mucha memoria guardar la información asi.
* No permite exportar el tipo de los datos

Por eso debemos exportarlo en el formato pickle para evitar estos dos inconvenientes.

In [ ]:
data.to_pickle("/content/drive/MyDrive/SisbenColombia/Sisben.pkl") 